src/system.ts

```diff
@@ -173,3 +188,86 @@ export const fetchCacheFromS3 = async () => {
     });
   }
 };
+
+export const saveCacheToS3 = async () => {
+  const startTimeMs = Date.now();
+
+  logger.info('Saving warp cache to S3', {
+    bucket,
+  });
+
+  try {
+    // read files from local file system
+    const parallelLimit = pLimit(10);
+    const uploadFolder = async ({
+      folderPath,
+      bucket,
+      keyPrefix,
+    }: {
+      folderPath: string;
+      bucket: string;
+      keyPrefix: string;
+    }) => {
+      const files = fs.readdirSync(folderPath);
+      await Promise.all(
+        files.map(async (file) => {
+          // wrap in a pLimit to avoid resource exhaustion
+          return parallelLimit(() => {
+            const filePath = path.join(folderPath, file);
+            if (fs.statSync(filePath).isFile()) {
+              logger.debug('Uploading file to S3', {
+                filePath,
+                bucket,
+                keyPrefix,
+              });
+              const fileStream = fs.createReadStream(filePath);
+              const key = path.basename(filePath);
+              const upload = new Upload({
+                client: s3,
+                params: {
+                  Bucket: bucket,
+                  Key: key,
+                  Body: fileStream,
+                },
+              });
+
+              // catch errors for a single file
+              return upload.done().catch((error: unknown) => {
+                const message =
+                  error instanceof Error ? error : new Error('Unknown error');
+                logger.error('Failed to upload file to S3', {
+                  error: message,
+                  file,
+                });
+              });
+            } else {
+              // recursively upload folders
+              return uploadFolder({
+                folderPath,
+                bucket,
+                keyPrefix: keyPrefix + file + '/',
+              });
+            }
+          });
+        }),
+      );
+    };
+
+    // upload files to S3 recursively and in a pLimit to avoid resource exhaustion
+    await uploadFolder({
+      folderPath: cacheDirectory,
+      bucket,
+      keyPrefix: '',
+    });
```

Comment:
will there be a cache limit to what this can recursively handle? 

it could be safer to run this in a `while (hasMoreFiles)` loop to avoid a huge stack

Reply:
warp shouldn't produce more than 10-15 files depending on our configuration - will test locally to confirm the recursion doesn't consume resources unexpectedly